### PR TITLE
feat: adds @oas.response decorator

### PR DESCRIPTION
**Pull request by mschnee**
Partially implements: #4300
See: #4406

## Adds
This PR adds the `@oas.response(code: number, ...modelOrSpec: ModelOrSpec[])` decorator.  

### Note
The decorator only affects the OpenAPI specification object that is created, and anything that uses it to verify and validate responses.  **This decorator has no impact on the actual response code**.  It also doesn't add any typings or lint configuration that could force response compliance at compile/lint time (if that's even possible).

### Example
```ts
@model()
class SuccessModel extends Model {
  constructor(err: Partial<SuccessModel>) {
    super(err);
  }
  @property({default: 'Hi there!'})
  message: string;
}

class FooNotFound extends Model {
  @property()
  message: string;
}

class BarNotFound extends Model {
  @property()
  message: string;
}

class BazNotFound extends Model {
  @property()
  message: string;
}

class MyController {
  @oas.get('/greet/{foo}/{bar}',
  @oas.response(200, SuccessModel)
  @oas.response(404, FooNotFound, BarNotFound)
  @oas.response(404, BazNotFound)
  greet() {
    return new SuccessModel({message: 'Hello, world!'});
  }
}
```

## Checklist
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "fs-extra": "^8.1.0",
     "husky": "^3.1.0",
     "lerna": "^3.20.2",
+    "open-cli": "^5.0.0",
```
Comment (raymondfeng) [inline package.json]: Please open another PR for `open-cli` related changes.

File: docs/site/decorators/Decorators_openapi.md [hunk 2]
```diff
@@ -427,6 +427,52 @@ export class SomeController {
 }
 ```
 
+#### anyOf, allOf, oneOf, not
```
Comment (raymondfeng) [inline docs/site/decorators/Decorators_openapi.md]: It would be nice to see another PR for this documentation. At lease a separate commit.

Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: This should have been in my `complex-x-ts-types` PR, I'll separate it out :)

File: package.json (hunk 1 above)
Comment (mschnee) [reply package.json]: :+1:

Sorry this keeps slipping in.  `npm run coverage` fails for me because the command `open coverage/index.html` is only valid on MacOS.  I'll make separate PRs.

**Comment by mschnee** [comment]
Note: I'd love to figure out some way to use these decorators as response code hints, but I'm not entirely sure how to go about doing that in a way that fits the current patterns.

e.g.
```
@oas.response(404, SomeErrorModel)
```

If the invoked method returns or throws `insranceof SomeErrorModel` (or ideally, a more direct test), and that model isn't associated with any other status codes for that route, and if the status has not already been set explicitly, then assume the developer intends to return 404.

File: docs/site/decorators/Decorators_openapi.md [hunk 3]
```diff
@@ -496,6 +496,185 @@ class MyOtherController {
...
+  // this is a valid ResponseObject
+  @oas.get('/reference-object',
+  @oas.response(200, { $ref: '#/path/to/schema' })
+  returnFromResponseObject() {
+    return {message: 'Hello, world!'};
+  }
+}
+```
+
+#### I need more options!
```
Comment (raymondfeng) [inline docs/site/decorators/Decorators_openapi.md]: `#### Using even more options`?

File: packages/openapi-v3/src/build-responses-from-metadata.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  });
+  return builder;
+}
+export function buildResponsesFromMetadata(
+  metadata: ResponseDecoratorMetadata,
+  existingOperation?: OperationObject,
+): OperationObject {
+  const builder = buildMapsFromMetadata(metadata);
+  const base = existingOperation
+    ? _.cloneDeep(existingOperation.responses)
```
Comment (raymondfeng) [inline packages/openapi-v3/src/build-responses-from-metadata.ts]: We have `cloneDeep` from `DecoratorFactory`.

Comment (mschnee) [reply packages/openapi-v3/src/build-responses-from-metadata.ts]: :+1:

File: docs/site/decorators/Decorators_openapi.md [hunk 5]
```diff
@@ -496,6 +496,187 @@ class MyOtherController {
...
+
+class GenericError extends Model {
+  @property()
+  message: string;
+}
+
+class MyController {
+  @oas.get('/greet',
+  @oas.response(200, SuccessModel)
+  @oas.response(500, SuccessModel)
```
Comment (raymondfeng) [inline docs/site/decorators/Decorators_openapi.md]: `SuccessModel` -> `GenericError`.

File: docs/site/decorators/Decorators_openapi.md [hunk 6]
```diff
@@ -496,6 +496,187 @@ class MyOtherController {
...
+  "paths": {
+    "/greet": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "Ok",
+            "content": {
+              "application/json": {
+                "schema": {
+                  "$ref": "#/components/schemas/GenericError"
```
Comment (raymondfeng) [inline docs/site/decorators/Decorators_openapi.md]: `#/components/schemas/SuccessModel`

File: packages/openapi-v3/src/types.ts [hunk 7]
```diff
@@ -32,3 +37,18 @@ export function createEmptyApiSpec(): OpenApiSpec {
 export interface TagsDecoratorMetadata {
   tags: string[];
 }
+
+export declare type ResponseModelOrSpec =
```
Comment (raymondfeng) [inline packages/openapi-v3/src/types.ts]: `export type`?

Comment (mschnee) [reply packages/openapi-v3/src/types.ts]: :+1:

File: docs/site/decorators/Decorators_openapi.md (hunk 6 above)
Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: :+1:

File: docs/site/decorators/Decorators_openapi.md (hunk 5 above)
Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: :+1:

File: docs/site/decorators/Decorators_openapi.md [hunk 8]
```diff
@@ -496,6 +496,187 @@ class MyOtherController {
 }
 ```
 
+### @oas.response
+
+[API document](https://loopback.io/doc/en/lb4/apidocs.openapi-v3.response.html),
```
Comment (raymondfeng) [inline docs/site/decorators/Decorators_openapi.md]: It will be https://loopback.io/doc/en/lb4/apidocs.openapi-v3.oas.response.html.

Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: :+1:

**Comment by raymondfeng** [comment]
@mschnee Can you rebase your PR against latest master branch with `git fetch --all && git rebase upstream/master`?

**Comment by mschnee** [comment]
Rebased from upstream and squashed!

File: docs/site/decorators/Decorators_openapi.md [hunk 9]
```diff
@@ -427,52 +427,6 @@ export class SomeController {
...
-        },
-      },
-    },
-  })
-  getSomeValue() {
-    return {foo: 'foo', bar: 'bar'};
-  }
-}
-```
-
```
Comment (raymondfeng) [inline docs/site/decorators/Decorators_openapi.md]: We should revert this. I think our intention is to keep this section for `anyOf, allOf, oneOf, not`.

Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: Yes, I see my other PR landed in so this should not be removed now.

File: docs/site/decorators/Decorators_openapi.md [hunk 10]
```diff
@@ -542,6 +496,187 @@ class MyOtherController {
...
+  message: string;
+}
+
+class GenericError extends Model {
+  @property()
+  message: string;
+}
+
+class MyController {
+  @oas.get('/greet',
```
[the remark this answers is not in the document]
Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: 👍

File: docs/site/decorators/Decorators_openapi.md [hunk 11]
```diff
@@ -542,6 +496,187 @@ class MyOtherController {
...
+  message: string;
+}
+
+class BazNotFound extends Model {
+  @property()
+  message: string;
+}
+
+class MyController {
+  @oas.get('/greet/{foo}/{bar}',
```
[the remark this answers is not in the document]
Comment (mschnee) [reply docs/site/decorators/Decorators_openapi.md]: Thanks for catching the typos!  They're the bane of my existence, and the reason I prefer to write compile-checked code like TypeScript - and why I wanted to create convenience decorators instead of manually maintaining json blobs ;)

File: package.json [hunk 12]
```diff
@@ -26,7 +26,6 @@
     "fs-extra": "^8.1.0",
     "husky": "^3.1.0",
     "lerna": "^3.20.2",
-    "open-cli": "^5.0.0",
```
Comment (raymondfeng) [inline package.json]: Please revert `open-cli` changes.

Comment (mschnee) [reply package.json]: :+1:

File: packages/openapi-v3/package.json [hunk 13]
```diff
@@ -20,6 +20,7 @@
     "@loopback/repository": "^1.19.1",
     "@loopback/testlab": "^1.10.3",
     "@types/debug": "^4.1.5",
+    "@types/http-status": "^1.1.2",
```
Comment (raymondfeng) [inline packages/openapi-v3/package.json]: I don't see `http-status` as a dev dependency.

File: packages/openapi-v3/src/__tests__/unit/decorators/response.decorator.unit.ts [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    class BarError extends Model {
+      constructor(err: Partial<BarError>) {
+        super(err);
+      }
+      @property({default: 'bar'})
+      bar: string;
+    }
+
+    const successSchema: ResponseObject = {
+      description: httpStatus['200'],
```
Comment (raymondfeng) [inline packages/openapi-v3/src/__tests__/unit/decorators/response.decorator.unit.ts]: This should require `http-status` to be a dev dependency. The build is passing because the module is transitively added as follows:
```
rest
└─┬ strong-error-handler@3.4.0
  └── http-status@1.4.2 
```

File: packages/openapi-v3/src/decorators/response.decorator.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+ *    @oas.response(404, OneError, { $ref: '#/definition...'})
+ *    @oas.response(403, SecondError, { schema: ... })
+ *  }
+ *
+ */
+export function response(
+  responseCode: number,
+  ...responseModelOrSpec: ResponseModelOrSpec[]
+) {
+  const messageKey = String(responseCode) as keyof httpStatus.HttpStatus;
```
Comment (raymondfeng) [inline packages/openapi-v3/src/decorators/response.decorator.ts]: This makes `http-status` a regular dependency (instead of dev). Please add it to package.json.

File: packages/openapi-v3/src/decorators/response.decorator.ts [hunk 16]
```diff
@@ -0,0 +1,83 @@
+// Copyright IBM Corp. 2018,2019. All Rights Reserved.
+// Node module: @loopback/openapi-v3
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import {MethodMultiDecoratorFactory} from '@loopback/core';
+import * as httpStatus from 'http-status';
```
Comment (raymondfeng) [inline packages/openapi-v3/src/decorators/response.decorator.ts]: It should be `import httpStatus from 'http-status'` now.

Comment (mschnee) [reply packages/openapi-v3/src/decorators/response.decorator.ts]: I'll make these changes today
